crates/holochain/src/core/sys_validate/tests.rs

```diff
@@ -476,135 +476,101 @@ async fn incoming_ops_filters_private_entry() {
 #[test]
 /// Test the chain validation works.
 fn valid_chain_test() {
-    let author = fixt!(AgentPubKey);
     // Create a valid chain.
-    let mut actions = vec![];
-    actions.push(ActionHashed::from_content_sync(Action::Dna(Dna {
-        author: author.clone(),
-        timestamp: Timestamp::from_micros(0),
-        hash: fixt!(DnaHash),
-    })));
-    actions.push(ActionHashed::from_content_sync(Action::Create(Create {
-        author: author.clone(),
-        timestamp: Timestamp::from_micros(1),
-        action_seq: 1,
-        prev_action: actions[0].to_hash(),
-        entry_type: fixt!(EntryType),
-        entry_hash: fixt!(EntryHash),
-        weight: Default::default(),
-    })));
-    actions.push(ActionHashed::from_content_sync(Action::Create(Create {
-        author: author.clone(),
-        timestamp: Timestamp::from_micros(2),
-        action_seq: 2,
-        prev_action: actions[1].to_hash(),
-        entry_type: fixt!(EntryType),
-        entry_hash: fixt!(EntryHash),
-        weight: Default::default(),
-    })));
+    let actions = vec![
+        TestChainItem::new(0),
+        TestChainItem::new(1),
+        TestChainItem::new(2),
+    ];
     // Valid chain passes.
     validate_chain(actions.iter(), &None).expect("Valid chain");
 
     // Create a forked chain.
     let mut fork = actions.clone();
-    fork.push(ActionHashed::from_content_sync(Action::Create(Create {
-        author: author.clone(),
-        timestamp: Timestamp::from_micros(10),
-        action_seq: 1,
-        prev_action: actions[0].to_hash(),
-        entry_type: fixt!(EntryType),
-        entry_hash: fixt!(EntryHash),
-        weight: Default::default(),
-    })));
+    fork.push(TestChainItem {
+        seq: 1,
+        hash: vec![111],
+        prev: Some(vec![0]),
+    });
```

Comment:
This seems to be the main benefit of this change. Namely, you can create test items with three fields instead of 7. What I don't have a good sense of, is how well does this translate to other testing scenarios?

Do we need sys validation to validate that the author remains consistent throughout the chain, or that the timestamps increase? i.e. are we going to need to just add all the fields back into your generic trait eventually?

Reply:
Yeah, this is the most notable benefit from looking at the diff, in terms of lines removed vs added. Some of the other benefits become more apparent when actually writing and reading failing tests though:

- When there's a failure, you see only signal in the output, no noise, especially when dealing with a sizeable list of fixtures.
- You can construct properties to test against directly rather than needing to refer to other pieces that were built up, i.e. `42` instead of `action1.hashed.seq`, so it's easier to see in the test what's actually being tested, and to match that to the output.

Tests only really come to life when they fail, so these benefits are for those cases.

Reply:
> are we going to need to just add all the fields back into your generic trait eventually?

I would imagine that each component of the system, as we start to test it more heavily, could have its own traits for the objects that it's working with. Sys validation at the highest levels may always just use concrete Actions, since the abstraction would just be a replication of the actual data, i.e., the signal-to-noise ratio is already quite high. But if there is a subcomponent of sys validation that we want to test heavily, it could be worth defining that component in terms of a trait of a subset of the full Action.

I focused on chain logic here because in my next feature I need to write a bunch more tests along these lines and want to be able to talk only about the "chaininess" of the actions. But other components could have their own abstractions for their own testing needs.

This speaks to the point about [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns). The `ChainItem` trait here is like a line in the sand, saying "this is all we need to know about to define a chain structure. If you are tempted to add something to this trait, be sure it actually has something to do with chain structure!". Basically we should really think hard about adding something to this trait -- the trait can actually help us be more disciplined in keeping our concerns separated. The trait tangibly represents a concern, and modifying it gives us visibility that our notion of that concern wants to expand, either in a desirable or undesirable way.

Reply:
This `ChainItem` is concerned with constructing a tree from a collection of actions, and also detecting whether that tree has multiple branches. It lets us define what "the chain" is, and what "forks" are.

Other chain validation sits on top of that: given that you have a chain, determine whether the author is the same, and whether timestamps increase monotonically, and other semantic things about the particular actions. But this fork detection / chain creation stuff can stay simple and unconcerned with that other stuff.